### PR TITLE
Optimze might_contain sparksql function

### PR DESCRIPTION
**Pull request by marin-ma**
Currently the bloom filter is deserialized and built per call of
BloomFilterMightContainFunction::apply. But in Spark application, it's
created as a constant value in each task thread in a stage. It can be
registered as a stateful function so that we can make it independently
registered in different tasks and stages, and initialize the bloom filter only
once in a task/Velox query thread.

We do this optimization because we have changed the default batch size from 
32k to 4k in Spark TPCH/TPCDS benchmark, and 4k gets better performance in 
TPCH workloads. But in TPCDS workloads we have seen queries like 
TPCDS Q24 has significant performance regression with 4k batch size. 
The perf hotspot shows that `facebook::velox::BloomFilter::merge` takes ~40%. 
This means smaller batch size leads to more iterations of calling 
`BloomFilterMightContain::apply` and the total cost of
deserializing and building the Bloom Filter is very heavy.

```
+   90.32%     0.12%  java             libvelox.so           [.] facebook::velox::exec::Driver::runInternal                                                  
+   89.84%     0.02%  java             libgluten.so          [.] Java_io_glutenproject_vectorized_ArrowOutIterator_nativeHasNext                             
+   89.77%     0.03%  java             libgluten.so          [.] gluten::ResultIterator::GetNext                                                             
+   89.76%     0.00%  java             libvelox.so           [.] arrow::Iterator<std::shared_ptr<gluten::ColumnarBatch> >::Next<gluten::ResultIterator::Wrapp
+   89.76%     0.00%  java             libvelox.so           [.] gluten::WholeStageResultIterator::Next                                                      
+   89.69%     0.01%  java             libvelox.so           [.] facebook::velox::exec::Task::next                                                           
+   89.61%     0.00%  java             libvelox.so           [.] facebook::velox::exec::Driver::next                                                         
+   72.24%     0.03%  java             libvelox.so           [.] facebook::velox::exec::FilterProject::getOutput                                             
+   71.84%     0.22%  java             libvelox.so           [.] facebook::velox::exec::Expr::eval                                                           
+   71.16%     0.12%  java             libvelox.so           [.] facebook::velox::exec::ConjunctExpr::evalSpecialForm                                        
+   71.12%     0.02%  java             libvelox.so           [.] facebook::velox::exec::ExprSet::eval                                                        
+   70.46%     0.02%  java             libvelox.so           [.] facebook::velox::exec::FilterProject::filter                                                
+   70.39%     0.03%  java             libvelox.so           [.] facebook::velox::exec::Expr::evalSpecialFormWithStats                                       
+   69.42%     0.04%  java             libvelox.so           [.] facebook::velox::exec::Expr::evalAllImpl                                                    
+   68.96%     0.07%  java             libvelox.so           [.] facebook::velox::exec::Expr::applyFunction                                                  
+   66.68%     3.63%  java             libvelox.so           [.] facebook::velox::functions::sparksql::(anonymous namespace)::BloomFilterMightContainFunction
+   44.87%    40.20%  java             libvelox.so           [.] facebook::velox::BloomFilter<facebook::velox::StlAllocator<unsigned long> >::merge          
+   40.21%     0.00%  java             libvelox.so           [.] 0x00007f86e3a78e90                                                                          
+   27.50%     0.00%  java             [JIT] tid 437053      [.] 0x00007f872361bfa3                                                                          
+   27.46%     0.00%  java             [JIT] tid 437053      [.] 0x00007f872406d737                                                                          
+   18.29%    18.29%  java             libc.so.6             [.] 0x00000000001a6cd5                                                                          
+   17.17%     0.04%  java             libvelox.so           [.] facebook::velox::exec::TableScan::getOutput                                                 
+   17.14%     0.00%  java             libvelox.so           [.] 0x00007f86e3a952c0                                                                          
+   16.96%     0.00%  java             libvelox.so           [.] std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construc
+   16.95%     0.09%  java             libvelox.so           [.] facebook::velox::connector::hive::HiveDataSource::next                                      
+   16.81%     0.04%  java             libvelox.so           [.] facebook::velox::parquet::ParquetRowReader::next                                            
+   15.75%     0.14%  java             libvelox.so           [.] facebook::velox::dwio::common::SelectiveStructColumnReaderBase::read                        
+   15.74%     0.02%  java             libvelox.so           [.] facebook::velox::dwio::common::SelectiveStructColumnReaderBase::next                        
+    8.17%     0.01%  java             libvelox.so           [.] facebook::velox::dwio::common::SelectiveIntegerColumnReader::readHelper<facebook::velox::par
```

After applying this fix, the runtime of TPCDS Q24a with SF2T decreased from 
69s to 24s on a 48C server.

### CODE REVIEW

**Comment by marin-ma** [comment]
cc @zhejiangxiaomai @jinchengchenghh

File: velox/functions/sparksql/MightContain.cpp [hunk 1]
```diff
@@ -33,26 +49,24 @@ class BloomFilterMightContainFunction final : public exec::VectorFunction {
...
-      auto contain = output.mayContain(
-          folly::hasher<int64_t>()(value->valueAt<int64_t>(row)));
-      result.set(row, contain);
-    });
   }
+
+ private:
+  std::shared_ptr<memory::MemoryPool> pool_;
+  HashStringAllocator allocator_;
+  BloomFilter<StlAllocator<uint64_t>> bloom_;
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/MightContain.cpp:68]: Please use bloomFilter_

File: velox/functions/sparksql/MightContain.cpp [hunk 2]
```diff
@@ -33,26 +49,24 @@ class BloomFilterMightContainFunction final : public exec::VectorFunction {
...
-    HashStringAllocator allocator{context.pool()};
-    VELOX_USER_CHECK(serialized->isConstantMapping())
-    BloomFilter output{StlAllocator<uint64_t>(&allocator)};
-    try {
-      auto sv = serialized->valueAt<StringView>(0);
-      output.merge(sv.data());
-    } catch (const std::exception& e) {
-      context.setErrors(rows, std::current_exception());
-      return;
+    if (!isSet_) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/MightContain.cpp:54]: Please use bloomFilter_.isSet()

File: velox/functions/sparksql/MightContain.cpp [hunk 3]
```diff
@@ -33,26 +49,24 @@ class BloomFilterMightContainFunction final : public exec::VectorFunction {
...
+      rows.applyToSelected([&](int row) { result.set(row, false); });
+    } else {
+      rows.applyToSelected([&](int row) {
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/MightContain.cpp:57]: Can we initialize the bloomFilter_ here? Then we can use context pool to track memory usage

File: velox/functions/sparksql/MightContain.cpp [hunk 4]
```diff
@@ -33,26 +49,24 @@ class BloomFilterMightContainFunction final : public exec::VectorFunction {
...
-    rows.applyToSelected([&](int row) {
-      auto contain = output.mayContain(
-          folly::hasher<int64_t>()(value->valueAt<int64_t>(row)));
-      result.set(row, contain);
-    });
   }
+
+ private:
+  std::shared_ptr<memory::MemoryPool> pool_;
+  HashStringAllocator allocator_;
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/MightContain.cpp:67]: We don't need to reserve allocator_, just bloomFilter_ is enough

File: velox/functions/sparksql/MightContain.cpp (hunk 2 above)
Comment (jinchengchenghh) [reply velox/functions/sparksql/MightContain.cpp:54]: what's the situation that it is not set?

File: velox/functions/sparksql/tests/MightContainTest.cpp [hunk 5]
```diff
@@ -25,17 +25,34 @@ namespace {
 class MightContainTest : public SparkFunctionBaseTest {
  protected:
   void testMightContain(
-      const VectorPtr& bloom,
+      const std::optional<std::string>& serialized,
```
Comment (jinchengchenghh) [inline velox/functions/sparksql/tests/MightContainTest.cpp:28]: This is VectorFunction, so we need to use input Vector, previously I use StringView as input, the final version changes it to vector

File: velox/functions/sparksql/MightContain.cpp (hunk 2 above)
Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp:54]: I think not set is for empty bloom filter. UT `nullBloomFilter` will run into this code path.

File: velox/functions/sparksql/MightContain.cpp (hunk 3 above)
Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp:57]: No, because this function is marked with `const`.

File: velox/functions/sparksql/MightContain.cpp (hunk 2 above)
Comment (jinchengchenghh) [reply velox/functions/sparksql/MightContain.cpp:54]: You can set bloomFilter_.isSet() to const, I have another PR to set it to const

File: velox/functions/sparksql/tests/MightContainTest.cpp (hunk 5 above)
Comment (marin-ma) [reply velox/functions/sparksql/tests/MightContainTest.cpp:28]: I think we don't need to pass a Vector here. Previously the first input argument is a vector of string, but now the serialized string will be added to the argument list as a single constant.

File: velox/functions/sparksql/MightContain.cpp (hunk 4 above)
Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp:66]: @mbasmanova Per discussion with Chengcheng, we use `HashStringAllocator` instead of the default allocator here is because we want to track the memory usage of serialized Bloom Filter in context memory pool. But if the bloom filter is initialized only once per thread, it should be a small piece of memory. I have tested that for 4k and 8k batch size, the `bits_.size()` in BloomFilter are both 128KB so it might be unnecessary to track using the context memory pool. Do you think we can remove the pool_ and allocator_ here and create the BloomFilter with default allocator?

Comment (mbasmanova) [reply velox/functions/sparksql/MightContain.cpp:66]: @marin-ma This sounds reasonable to me.

**Comment by marin-ma** [comment]
@mbasmanova Can you help review this PR?

File: velox/functions/sparksql/MightContain.cpp [hunk 6]
```diff
@@ -15,14 +15,21 @@
...
+#include <utility>
+
 #include "velox/common/base/BloomFilter.h"
 #include "velox/common/memory/HashStringAllocator.h"
 #include "velox/expression/DecodedArgs.h"
 #include "velox/vector/FlatVector.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
+template <typename Allocator = std::allocator<uint64_t>>
```
Comment (mbasmanova) [inline velox/functions/sparksql/MightContain.cpp:27]: Let's remote template parameter since it is always the same.

File: velox/functions/sparksql/MightContain.cpp [hunk 7]
```diff
@@ -64,8 +66,22 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> mightContainSignatures() {
...
 
-std::unique_ptr<exec::VectorFunction> makeMightContain() {
-  return std::make_unique<BloomFilterMightContainFunction>();
+std::shared_ptr<exec::VectorFunction> makeMightContain(
+    const std::string& name,
+    const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_CHECK_EQ(inputArgs.size(), 2);
+  BaseVector* serialized = inputArgs[0].constantValue.get();
+  if (!serialized) {
+    VELOX_USER_FAIL(
```
Comment (mbasmanova) [inline velox/functions/sparksql/MightContain.cpp:75]: user VELOX_USER_CHECK_NOT_NULL(serialized, ...)

File: velox/functions/sparksql/MightContain.cpp [hunk 8]
```diff
@@ -64,8 +66,22 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> mightContainSignatures() {
...
+        "{} requires first argument to be a constant of type VARBINARY", name);
+  }
+  BloomFilter bloomFilter;
+  if (!serialized->isNullAt(0)) {
```
Comment (mbasmanova) [inline velox/functions/sparksql/MightContain.cpp:79]: Can this be null? I assume no. Let's make this a VELOX_CHECK

File: velox/functions/sparksql/tests/MightContainTest.cpp [hunk 9]
```diff
@@ -25,17 +25,34 @@ namespace {
...
       const VectorPtr& value,
       const VectorPtr& expected) {
-    auto result = evaluate(
-        "might_contain(cast(c0 as varbinary), c1)",
-        makeRowVector({bloom, value}));
-    velox::test::assertEqualVectors(expected, result);
+    // Not using `evaluate()` because the bloom filter binary cannot be parsed
+    // by DuckDB parser.
+    auto selected = SelectivityVector(value->size());
+    std::vector<std::shared_ptr<const velox::core::ITypedExpr>> args;
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/MightContainTest.cpp:34]: use TypedExprPtr

**Review by mbasmanova — COMMENTED** [review]
@marin-ma Thank you for writing detailed PR description. Would you please wrap it to keep each line <= 80 characters. Some minor comments below.

File: velox/functions/sparksql/MightContain.cpp (hunk 8 above)
Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp:79]: @jinchengchenghh Can you help to answer this question? If we add this check the test "nullBloomFilter" in MightContainTest will fail.

Comment (jinchengchenghh) [reply velox/functions/sparksql/MightContain.cpp:79]: Spark might_contain return null if any of input argument is null.
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/BloomFilterAggregateQuerySuite.scala#L363

Previously, we use `isDefaultNullBehavior = true` to treat null inputs, but now, it cannot take effect.
So we need to check if it is null.

Comment (mbasmanova) [reply velox/functions/sparksql/MightContain.cpp:79]: Are you saying that the framework ends up creating function instance even when there is a constant input that is null? Looks like a bug to me. Would you file GitHub issue about it? CC: @laithsakka @Yuhta @bikramSingh91 

For now then, let's return an instance of AlwaysFailingVectorFunction when input is null.

File: velox/functions/sparksql/MightContain.cpp [hunk 10]
```diff
@@ -23,6 +25,11 @@
 namespace facebook::velox::functions::sparksql {
 namespace {
```
Comment (laithsakka) [inline velox/functions/sparksql/MightContain.cpp:26]: why is this not implemented as simple function?

File: velox/functions/sparksql/MightContain.cpp (hunk 3 above)
Comment (laithsakka) [reply velox/functions/sparksql/MightContain.cpp:48]: the function being const should not be a limitation you can always mark bloomFilter_ as mutable

File: velox/functions/sparksql/MightContain.cpp (hunk 8 above)
Comment (laithsakka) [reply velox/functions/sparksql/MightContain.cpp:79]: I am a bit lost here, why shall we create AlwaysFailingVectorFunction if expected behaviour is to return null?

Comment (jinchengchenghh) [reply velox/functions/sparksql/MightContain.cpp]: @mbasmanova  No, it's normal to check null when creating function instance because input may be null.

Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp]: @mbasmanova I agree with Chengcheng. https://github.com/apache/spark/blob/b496edb5d476b3ec13b5c69f3ab58b81e603e212/sql/core/src/test/scala/org/apache/spark/sql/BloomFilterAggregateQuerySuite.scala#L358-L372 In Spark's UT the bloom_filter_agg can produce null output with empty input, and in this case the input of "might_contain" is null. Here, the "might_contain" need to keep the same behavior as a sparksql function. I think this commit should be reverted https://github.com/facebookincubator/velox/pull/4728/commits/c12561d5294e21f810e977bc92cda768d8e25475

**Comment by marin-ma** [comment]
@mbasmanova Do you have any other comments? Should I revert this commit https://github.com/facebookincubator/velox/commit/c12561d5294e21f810e977bc92cda768d8e25475 to keep the same null behavior as Spark?

File: velox/expression/VectorFunction.h [hunk 11]
```diff
@@ -137,6 +137,10 @@ class AlwaysFailingVectorFunction final : public VectorFunction {
   explicit AlwaysFailingVectorFunction(std::exception_ptr exceptionPtr)
       : exceptionPtr_{std::move(exceptionPtr)} {}
 
+  bool isDefaultNullBehavior() const override {
```
Comment (mbasmanova) [inline velox/expression/VectorFunction.h:140]: Why this change? It doesn't look right.

File: velox/functions/sparksql/MightContain.cpp (hunk 8 above)
Comment (mbasmanova) [reply velox/functions/sparksql/MightContain.cpp]: @marin-ma Thank you for clarifying. I agree that we should make sure the semantics are identical to Spark's.

File: velox/expression/VectorFunction.h (hunk 11 above)
Comment (marin-ma) [reply velox/expression/VectorFunction.h:140]: When the input arg is null, the exception in AlwaysFailingVectorFunction::apply won't be thrown if it's "true".

File: velox/functions/sparksql/MightContain.cpp (hunk 10 above)
Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp:26]: The `bloomFilter_` can be different for each execution thread, but is a constant within a thread. From the document it should be registered as a stateful function. https://github.com/facebookincubator/velox/blob/6b38117f9bbe89703888d949e1121ac5e394e968/velox/docs/develop/scalar-functions.rst?plain=1#L1138-L1140

**Comment by marin-ma** [comment]
@mbasmanova @laithsakka Do you have further comments? Can this one be merged?

**Comment by marin-ma** [comment]
@mbasmanova Do you have any other comments or feedbacks? I see all GHA passed.

File: velox/functions/sparksql/MightContain.cpp (hunk 10 above)
Comment (mbasmanova) [reply velox/functions/sparksql/MightContain.cpp:26]: @marin-ma I believe Laith is suggesting that this can be implemented as a simple function and bloom filter can be initialized in 'initialize' method.

See https://facebookincubator.github.io/velox/develop/scalar-functions.html#access-to-session-properties-and-constant-inputs

Comment (marin-ma) [reply velox/functions/sparksql/MightContain.cpp]: @mbasmanova @laithsakka Updated the code to use SimpleFunction in the latest commit, but the function signature changed to non-const, which caused the CI to fail. Is there any way to register a const argument?
```
Found differences: {'iterable_item_added': {"root['might_contain'][0]": '(varbinary,bigint) -> boolean'}, 'iterable_item_removed': {"root['might_contain'][0]": '(constant varbinary,bigint) -> boolean'}}
 
            Incompatible changes in function signatures have been detected.
            This means your changes have modified function signatures and possibly broken backwards compatibility. 
```

Comment (mbasmanova) [reply velox/functions/sparksql/MightContain.cpp]: @kgpai Krishna, would you help look into this?

Comment (kgpai) [reply velox/functions/sparksql/MightContain.cpp]: @mbasmanova Should we remove the constant attribute from signatures ? I am hesitant to do that in general - Let me know though if its not required in function signatures .

Comment (laithsakka) [reply velox/functions/sparksql/MightContain.cpp]: @marin-ma  Simple function interface does not support marking args as constant However its a simple change.
, would it make sense to by pass that failure (aprove the signature change) until I add that support i can ping you then to add that part?

File: velox/functions/sparksql/MightContain.h [hunk 12]
```diff
@@ -13,12 +13,36 @@
...
 
-std::vector<std::shared_ptr<exec::FunctionSignature>> mightContainSignatures();
+template <typename T>
+struct BloomFilterMightContainFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
 
-std::unique_ptr<exec::VectorFunction> makeMightContain();
+  using Allocator = std::allocator<uint64_t>;
+
+  FOLLY_ALWAYS_INLINE void initialize(
```
Comment (laithsakka) [inline velox/functions/sparksql/MightContain.h:28]: no need for FOLLY_ALWAYS_INLINE this is not on a hot path

**Review by laithsakka — COMMENTED** [review]
Function looks neat if the circleCI  The signature change concern is not concerning me, 
I will work on supporting marking args as constant in the simple function interface and ping you

**Review by laithsakka — COMMENTED** [review]
Function looks neat if the circleCI  The signature change concern is not concerning me, 
I will work on supporting marking args as constant in the simple function interface and ping you

File: velox/functions/sparksql/MightContain.cpp (hunk 10 above)
Comment (kgpai) [reply velox/functions/sparksql/MightContain.cpp]: @laithsakka @mbasmanova In this case I think we can ignore this failure as a legitimate case. Let me know if you think otherwise.

**Review by laithsakka — APPROVED** [review]

**Review by laithsakka — APPROVED** [review]
